pkg/driver/node.go

```diff
@@ -118,37 +119,62 @@ func (d *nodeService) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 	}
 
 	klog.V(5).Infof("NodePublishVolume: mounting juicefs with secret %+v, options %v", reflect.ValueOf(secrets).MapKeys(), mountOptions)
-	jfs, err := d.juicefs.JfsMount(secrets, mountOptions)
+	jfs, err := d.juicefs.JfsMount(volumeID, secrets, mountOptions)
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "Could not mount juicefs: %v", err)
 	}
 
-	bindSource := jfs.GetBasePath()
+	var bindSource string
+	if m, ok := req.GetVolumeContext()["mode"]; ok && m == "dynamic" {
+		// dynamic provisioning
+		var capacity int64 = juicefs.DefaultCapacityBytes
+		if c, ok := req.GetVolumeContext()["capacity"]; ok {
+			if v, err := strconv.ParseInt(c, 10, 64); err == nil {
+				capacity = v
+			}
+		}
+
+		vol, err := jfs.CreateVol(volumeID, capacity)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not create volume: %s, %v", volumeID, err)
+		}
+
+		bindSource = vol.VolumePath
+	} else {
+		// static provisioning
+		bindSource = jfs.GetBasePath()
+	}
 	if subPath, ok := req.GetVolumeContext()["subPath"]; ok {
 		bindSource = path.Join(bindSource, subPath)
 	}
+
 	klog.V(5).Infof("NodePublishVolume: binding %s at %s with options %v", bindSource, target, mountOptions)
 	if err := d.juicefs.Mount(bindSource, target, fsTypeNone, []string{"bind"}); err != nil {
 		os.Remove(target)
 		return nil, status.Errorf(codes.Internal, "Could not bind %q at %q: %v", bindSource, target, err)
 	}
 
-	klog.V(5).Infof("NodePublishVolume: mounted %s at %s with options %v", source, target, mountOptions)
+	klog.V(5).Infof("NodePublishVolume: mounted %s at %s with options %v", volumeID, target, mountOptions)
 	return &csi.NodePublishVolumeResponse{}, nil
 }
 
 // NodeUnpublishVolume is a reverse operation of NodePublishVolume. This RPC is typically called by the CO when the workload using the volume is being moved to a different node, or all the workload using the volume on a node has finished.
 func (d *nodeService) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpublishVolumeRequest) (*csi.NodeUnpublishVolumeResponse, error) {
-	klog.V(4).Infof("NodeUnpublishVolume: called with args %+v", req)
+	// klog.V(4).Infof("NodeUnpublishVolume: called with args %+v", req)
 
 	target := req.GetTargetPath()
 	if len(target) == 0 {
 		return nil, status.Error(codes.InvalidArgument, "Target path not provided")
 	}
 
+	klog.V(5).Infof("NodeUnpublishVolume: unmounting ref for target %s", target)
+	if meta, err := juicefs.ReadMetaFile(target); err == nil {
```

Comment:
mount point and volume path are stored in meta file. this seems to be the only way to get mount point from `target`, since we bind a sub directory inside JFS to `target`, `GetMountRefs` can not find the correct mount point, so meta file is needed here, to cleanup JFS root mount point. otherwise there's gonna be a leak of jfsmount processes.